### ISSUE TITLE
Unable to authenticate

### ISSUE THREAD

**Issue by jonahadmin** [StackStorm/st2-auth-backend-ldap#12]
I have installed a default installation of Stackstorm v2.3 on Centos 7.2 and for the life of me I can't get LDAP authentication to work.  I've tried countless configurations and I'm now trying to do the minimum possible just to get it to authenticate against *any* account, to no avail.

This is what I have in my auth section:
```
[auth]
mode = standalone
backend = ldap
backend_kwargs ={"ldap_uri": "ldap://activedirectoryserver.XXX.com","use_tls": true,"bind_dn": "sAMAccountName={username}@XXX.com","bind_pw": "{password}","user": {"base_dn": "dc=XXX,dc=com","search_filter": "(sAMAccountName={username})","scope": "subtree"}} 
use_ssl = false
debug = true
enable = True
cert=/etc/ssl/st2/XXX.com-2017/combined.pem
key=/etc/ssl/st2/XXX.com-2017/privatekey.pem
logging = /etc/st2/logging.auth.conf
```

Below is what I see in the log:
```[root@it-ops01 bin]# ./st2auth --config-file /etc/st2/st2.conf
2017-06-21 18:11:18,234 DEBUG [-] Using config files: /etc/st2/st2.conf
2017-06-21 18:11:18,234 DEBUG [-] Using logging config: /etc/st2/logging.auth.conf
2017-06-21 18:11:18,238 INFO [-] Connecting to database "st2" @ "0.0.0.0:27017" as user "None".
2017-06-21 18:11:18,423 INFO [-] ST2 Auth API running in "standalone" auth mode
2017-06-21 18:11:18,423 INFO [-] (PID=1416) ST2 Auth API is serving on http://0.0.0.0:9100.
2017-06-21 18:11:18,423 INFO [-] Creating st2auth: StackStorm v2.3.0 as OpenAPI app.
2017-06-21 18:11:19,808 INFO [-] (1416) wsgi starting up on http://0.0.0.0:9100
2017-06-21 18:11:24,575 INFO [-] ad6289a0-ee8f-40d5-aae1-1a931aa3299c - POST /tokens with query={} (remote_addr='127.0.0.1',method='POST',request_id='ad6289a0-ee8f-40d5-aae1-1a931aa3299c',query={},path='/tokens')
2017-06-21 18:11:24,880 AUDIT [-] Invalid credentials provided (remote_addr='127.0.0.1',auth_backend='LDAPAuthenticationBackend')
2017-06-21 18:11:24,881 ERROR [-] Failed to call controller function "post" for operation "st2auth.controllers.v1.auth:token_controller.post": Invalid or missing credentials
Traceback (most recent call last):
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/router.py", line 414, in __call__
    resp = func(**kw)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2auth/controllers/v1/auth.py", line 78, in post
    **kwargs)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2auth/handlers.py", line 215, in handle_auth
    abort_request()
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2auth/handlers.py", line 37, in abort_request
    return abort(status_code, message)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/router.py", line 52, in abort
    raise exc.status_map[status_code](message)
HTTPUnauthorized: Invalid or missing credentials
2017-06-21 18:11:24,883 INFO [-] ad6289a0-ee8f-40d5-aae1-1a931aa3299c - 401 55 308.343ms (content_length=55,request_id='ad6289a0-ee8f-40d5-aae1-1a931aa3299c',runtime=308.343,remote_addr='127.0.0.1',status=401,method='POST',path='/tokens')```

I can't tell if it's even trying to connect to the LDAP server.  When I have use_ssl=true, then I am not able to reach st2auth at all, with the only error that nginx couldn't connect to st2auth, which makes me wonder if there's a more fundamental problem.

I'm completely at a loss now as to how to proceed.

**Comment by nzlosh** [comment]
Sorry to hear you're having a hard time.  To enable debugging, add `debug = True` to the `[system]` section of `st2.conf` configuration file.

If that doesn't give you the clue you need, post the debug output (be careful not to post sensitive information) and someone may be able to help.

I see you're using ActiveDirectory, there have been others that had to adapt the filter query to limit it to a subtree.  There may be hints to help here https://github.com/StackStorm/st2-auth-backend-ldap/issues/6

**Comment by jonahadmin** [comment]
Sorry for taking so long.  Something came up and I wasn't able to revisit this until now.

Starting from the example you provided, I was able to at least get it work.  The issue appears to be documentation.
1.  The "Use TLS" parameter is incorrectly documented.  If you do not want to Use TLS, you must omit the parameter entirely.  The mere existence of the parameter causes TLS to be turned on.  This is unrelated to the issue I was having, but was something I noticed in my experiments.
2. The Active Directory example is incorrect.  The bind_dn parameter should only be "{username}@example.com", not "sAMAccountName={username}@example.com".

I have not played around with the subtree settings yet.  Now that I have a working starting point, I will play around further with the settings.

**Comment by jonahadmin** [comment]
Okay, so I have a working configuration, using the following parameters for backend_kwargs:
backend_kwargs = {
	"ldap_uri": "ldaps://controller.exsmple.com", 
	"bind_dn": "{username}@example.com", 
	"bind_pw": "{password}",
	"user":{ 
		"base_dn": "OU=myorgunit,DC=example,DC=com",
		"scope": "subtree",
		"search_filter": "(&(sAMAccountName={username})(!(userAccountControl:1.2.840.113556.1.4.803:=2)))" 
		}
	}

Notably, I added an additional filter criteria that only matches users who are _not_ disabled in Active Directory.

Like the user in #6, subtree searching appears to be broken for me as well, so I am stuck within the base_dn for user account checking.  Additionally, the group criteria portion doesn't work at all for Active Directory, as it uses the member attribute and the value is a user's DN.  I tried using {dn} in the group filter but it didn't appear to work, so I don't appear to have a way of restricting access except to users within a specific OU.
